src/cargo/core/shell.rs

```diff
@@ -81,9 +81,15 @@ enum ShellOut {
     /// A plain write object without color support
     Write(Box<dyn Write>),
     /// Color-enabled stdio, with information on whether color should be used
+    ///
+    /// The separate buffered fields are used for buffered writing to the
+    /// corresponding stream. The non-buffered fields should be used when you
+    /// do not want content to be buffered.
     Stream {
         stdout: StandardStream,
+        buffered_stdout: BufferWriter,
```

Comment:
Is it possible for us to use [`termcolor::BufferedStandardStream`](https://docs.rs/termcolor/latest/termcolor/struct.BufferedStandardStream.html) here?

Reply:
I don't think so, at least not without auditing and modifying all users of the streams. The way BufferedStandardStream works is that all writes to it go to a buffer, and the buffer isn't flushed unless there is a manual call to flush, or the `Shell` gets dropped. I would like to ensure that all output is sent immediately in case cargo itself gets terminated (which would cause the output to be otherwise lost if it was buffered). I'm not confident we can have the necessary discipline to make sure things get flushed in all the right places (especially where callers access the stream directly like with `out()` and `err()`). I didn't see a way to get unbuffered access to a `BufferedStandardStream`. (In theory, we could write a wrapper that implements `Write` and always flushes, but that seems like too much complexity.)